### PR TITLE
Calculated travel time to GPs, pharmacies, and sports facilities

### PR DESCRIPTION
**Pull request by matthewgthomas**

### CODE REVIEW

File: data-raw/healthy-places/travel-time-to-gp.R [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+# Get Local Authorities in Great Britain for the next step
+ni_lad <-
+  boundaries_ltla21 |>
+  filter(str_detect(ltla21_code, "^N"))
+
+# Some GPs are not in Northern Ireland - remove them from the dataset
+ni_GPs <-
+  GPs$osm_points[ni_lad, ] |>
+  select(osm_id)
```
Comment (MikeJohnPage) [inline data-raw/healthy-places/travel-time-to-gp.R:31]: The `GPs` dataset also contains 77 polygons and 2 multipolygons. I scanned the osmdata package docs and tried pritning these polygons with `ggplot(GPs$osm_polygons) + geom_sf(colour = "black", fill = NA)`, but can't figure out what they are.

Are you sure they can be dropped?

File: data-raw/healthy-places/travel-time-to-gp.R [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+neighbours <- st_touches(ni_lad)
+
+# Look up the Local Authority each GP is in
+ni_GPs_lad <-
+  st_join(ni_GPs, ni_lad) |>
+  filter(!is.na(ltla21_code)) |>
+  mutate(
+    # Round coords to 3 decimal points to save memory
+    lat = st_coordinates(geometry)[,2] |> round(3),
+    lng = st_coordinates(geometry)[,1] |> round(3)
```
Comment (MikeJohnPage) [inline data-raw/healthy-places/travel-time-to-gp.R:55]: Are you sure the rounding is saving memory? I ran some quick tests with `lobstr::obj_size()` and whether the rounding is or isn't included, `ni_GPs_lad` comes out to 75.93 kB in any case.

File: data-raw/healthy-places/travel-time-to-gp.R [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  ) |>
+  st_drop_geometry() |>
+  select(
+    sdz21_code = SDZ2021_code,
+    lat,
+    lng
+  )
+
+# ---- Calculate travel time between data zones and GPs ----
```
Comment (MikeJohnPage) [inline data-raw/healthy-places/travel-time-to-gp.R:91]: As I don't have traveltime credentials, I can't run this code to check the working. But, scanning the code, the logic looks sound.

I do however, recommend refactoring this code. It uses a for loop nested within a for loop, with `GP_travel_time` not being preallocated a size before the loop. As far as I am aware, for loops without preallocation can be slow due to dynamic memory reallocation. When you don't preallocate the output vector, R needs to repeatedly resize the vector, which involves copying the entire data structure each iteration. This results in slow code. Additionally, the logic is quite hard to follow (and will be hard to debug).

I recommend:
- splitting the logic into functions
- using `purrr::map()` and family in place of the for loops (or you can use base `Map()` or `lapply()`)

File: data-raw/healthy-places/travel-time-to-gp.R [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  left_join(lookup_sdz_lad)
+
+# What are the fastest travel times within each Intermediate Zone?
+GP_travel_time_fastest <-
+  GP_travel_time |>
+  select(-osm_id) |>  # We don't need to know the GP ID for this
+  group_by(sdz21_code) |>
+  filter(travel_time_mins == min(travel_time_mins)) |>
+  ungroup() |>
+  distinct()
```
Comment (MikeJohnPage) [inline data-raw/healthy-places/travel-time-to-gp.R:212]: Why do we calculate/use the fastest travel time here. Wouldn't something like the mean be better:
```
GP_travel_time |>
  select(-osm_id) |> 
  summarize(travel_time_mins = mean(travel_time_mins), .by = "sdz21_code")
```
Or alternatively, the slowest travel time, so those most in need are covered by the analysis?

**Review by MikeJohnPage — CHANGES_REQUESTED** [review]
Given that the GP, pharmacies and sport centre scripts use almost identical logic, it would make a lot of sense to extract this code out into functions so that should a change need to be made, it can be made once.

File: data-raw/healthy-places/travel-time-to-gp.R (hunk 1 above)
Comment (matthewgthomas) [reply data-raw/healthy-places/travel-time-to-gp.R:31]: Yes - those points are not in Northern Ireland.

This code plots their locations:

```r
ni_disjoint <- 
  GPs$osm_points[ni_lad, , op = st_disjoint] |>
  select(osm_id)

# - Test plot to check GP locations -
ni_lad |>
  ggplot() +
  geom_sf(
    fill = NA,
    colour = "black"
  ) +
  geom_sf(
    data = ni_disjoint,
    inherit.aes = FALSE,
    colour = "red"
  ) +
  geom_sf(
    data = ni_GPs,
    inherit.aes = FALSE,
    colour = "blue"
  )
```

File: data-raw/healthy-places/travel-time-to-gp.R (hunk 2 above)
Comment (matthewgthomas) [reply data-raw/healthy-places/travel-time-to-gp.R:55]: That's very helpful to know. We can remove the rounding in that case.

File: data-raw/healthy-places/travel-time-to-gp.R (hunk 3 above)
Comment (matthewgthomas) [reply data-raw/healthy-places/travel-time-to-gp.R:91]: That's a great idea. Let's explore this next year. The need to `Sys.sleep(2)` slows the loop down anyway, so there will be a limit to how fast it can run. I'm not a strong enough functional programmer to be able to convert the for loops while making sure that running `time_filter_fast(...)` via `purrr:map()` doesn't exceed our API limits.

File: data-raw/healthy-places/travel-time-to-gp.R (hunk 4 above)
Comment (matthewgthomas) [reply data-raw/healthy-places/travel-time-to-gp.R:212]: This is because the England Health Index [calculates average minimum distance](https://www.ons.gov.uk/peoplepopulationandcommunity/healthandsocialcare/healthandwellbeing/methodologies/healthindexindicatorsanddefinitions#healthy-places-access-to-services) to travel to the nearest GP/pharmacy/sports centre. Replicating that here, but with travel time instead of distance.

Comment (MikeJohnPage) [reply data-raw/healthy-places/travel-time-to-gp.R:212]: I find the term "**average** minimum" unclear here. How can something be both an average and a minimum? Do we have access to any code or maths which shows how they calculate this?

I feel like the approach taken in this index isn't quite right. By taking the minimum, is this metric not sensitive to highly skewed left tails on the distribution of travel times? What do you think instead about taking the first quartile mean?

Comment (matthewgthomas) [reply data-raw/healthy-places/travel-time-to-gp.R:212]: I'm reading "average minimum" as the average of the minimum travel times/distances for the neighbourhoods within a Local Authority. But appreciate it's unclear. As far as I'm aware, there aren't any detailed technical docs nor code published anywhere.

I take your point about skewed tails. I'll explore the data a bit more this week and have a think about it.

Comment (matthewgthomas) [reply data-raw/healthy-places/travel-time-to-gp.R:212]: Looking through the [other indicators](https://www.ons.gov.uk/peoplepopulationandcommunity/healthandsocialcare/healthandwellbeing/methodologies/healthindexindicatorsanddefinitions) included in the Health Index, many of them are averages (e.g. crude death rate, percentage of people with (physical health condition), crude crime rates, etc.). For consistency, I'm now wondering whether we should take the mean travel time from neighbourhoods (Super Data Zones, in this case) in each Local Authority. What do you think @MikeJohnPage ?

The travel times from each Super Data Zone to GPs, pharmacies and sports centres are relatively normally distributed:
image

Comment (matthewgthomas) [reply data-raw/healthy-places/travel-time-to-gp.R]: That's right - these are all the travel times from the centroid of each super data zone within each local authority. We don't know how travel times are distributed within a super data zone because we don't have any more spatially granular data (e.g. the equivalents of output areas).

**Comment by matthewgthomas** [comment]
@MikeJohnPage I've updated the code to calculate mean travel times. Once we merge these changes, I'll change the Scotland Health Index.
